File: Medium/2316_Count_Unreachable_Pairs_of_Nodes_in_an_Undirected_Graph.py
class Solution:
    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        
        adj = [[] for i in range(n)]
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])

        ans = 0
        vis = [False] * n

        def bfs(src):
            count = 1
            vis[src] = True
            q = deque([src])

            while q:
                u = q.popleft()
                for v in adj[u]:
                    if not vis[v]:
                        vis[v] = True
                        count += 1
                        q.append(v)
            return count
        
        for i in range(n):
            if not vis[i]:
                cnt = bfs(i)
                ans += (n - cnt) * cnt
        return ans // 2

File: Medium/2316_Count_Unreachable_Pairs_of_Nodes_in_an_Undirected_Graph.cpp
class Solution {
public:

    long long bfs(vector<vector<int>> &adj, int src, vector<bool> &vis)
    {
        long long count = 1;
        vis[src] = true;
        queue<int> q;
        q.push(src);

        while(!q.empty())
        {
            int u = q.front();
            q.pop();

            for(auto &v: adj[u])
            {
                if(!vis[v])
                {
                    vis[v] = true;
                    count++;
                    q.push(v);
                }
            }
        }
        // cout << count << " ";
        return count;
    }
    long long countPairs(int n, vector<vector<int>>& edges) {
        
        vector<vector<int>> adj(n, vector<int> ());
        for(auto &edge: edges)
        {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        long long ans = 0;
        vector<bool> vis(n, false);
        for(int i=0 ; i<n ; i++)
        {
            if(!vis[i])
            {
                long long cnt = bfs(adj, i, vis);
                ans += (n - cnt) * cnt;
            }
        }
        return ans/2;
    }
};
